File: backend/src/scheduler/core/slot-score.ts
import {
  ceilToSlot,
  floorToSlot,
  Interval,
  MS_PER_MINUTE,
  overlapsAny,
  SLOT_MS,
} from "./slot";
import { effectivePreferenceMatrix, preferenceScoreAt } from "./preference";
import { utcToMinutes } from "../../common/utils";
import { STABILITY_SATURATION_HOURS, STABILITY_WEIGHT } from "../constants";

const HOUR_MS = 60 * MS_PER_MINUTE;

/**
 * Slot preference scoring + best-free-slot search — pure, no I/O, no clock,
 * no randomness (CLAUDE.md invariant 2). Matrix helpers live in
 * `preference.ts`; interval math in `slot.ts`.
 */

/**
 * Overlap-weighted preference score of a concrete interval `[startMs, endMs)`:
 * the sum, over every local clock-hour block `[h, h+1)` the interval touches,
 * of `overlapFraction · pref[weekday(h)][h]`, where `overlapFraction` is the
 * portion of that hour block covered by the interval (0…1). A slot that only
 * partially covers an hour contributes that hour fractionally — e.g. a
 * 09:15–11:00 slot scores `0.75·pref[..][9] + 1.0·pref[..][10]`. `[start, end)`
 * is half-open, so the block starting exactly at `end` is never scored.
 * `prefMatrix` is passed through {@link effectivePreferenceMatrix}. Pure — the
 * local hour boundaries are derived from `timezone` (handles fractional UTC
 * offsets and DST).
 */
export function slotPreferenceScore(
  prefMatrix: number[],
  startMs: number,
  endMs: number,
  timezone: string,
): number {
  const matrix = effectivePreferenceMatrix(prefMatrix);
  let total = 0;
  let cursor = startMs;
  while (cursor < endMs) {
    // End of the local clock-hour block containing `cursor`.
    const minuteOfHour = utcToMinutes(new Date(cursor), timezone) % 60;
    const blockEnd = cursor + (60 - minuteOfHour) * MS_PER_MINUTE;
    const segEnd = Math.min(blockEnd, endMs);
    const weight = (segEnd - cursor) / HOUR_MS;
    total += weight * preferenceScoreAt(matrix, new Date(cursor), timezone);
    cursor = segEnd;
  }
  return total;
}

/**
 * Signed, ready-to-add "don't churn this without good reason" penalty for
 * moving a session away from the start time the user last set manually
 * (`prevStartMs`) — same "return the final signed contribution" convention
 * as {@link slotPreferenceScore}, so callers just add it directly. Always
 * `≤ 0`: the further `newStartMs` sits from `prevStartMs`, the more negative,
 * saturating at `STABILITY_SATURATION_HOURS` so a candidate a day away is no
 * more penalized than one just past the saturation point — see
 * `STABILITY_WEIGHT` in `constants.ts` for why the cap is sized the way it
 * is relative to the other scoring terms.
 */
export function stabilityScore(prevStartMs: number, newStartMs: number) {
  const distanceHours = Math.abs(newStartMs - prevStartMs) / HOUR_MS;
  const saturation =
    Math.min(distanceHours, STABILITY_SATURATION_HOURS) /
    STABILITY_SATURATION_HOURS;
  return -STABILITY_WEIGHT * saturation;
}

/**
 * Scan every 15-minute-aligned candidate start in `[windowStart, windowEnd)`
 * and return the free one (no overlap with `occupied`) whose start-hour has the
 * highest preference score. Earliest start wins ties. `null` when nothing free
 * fits.
 *
 * `windowEnd` bounds the latest legal **start**. `fitWindowEnd` (default
 * `windowEnd`) bounds the latest legal **end** — pass a later instant to let a
 * task start before `windowEnd` yet finish after it. Cross-midnight placement
 * uses `windowEnd = nextMidnight`, `fitWindowEnd = min(deadline, nextMidnight +
 * duration − one slot)`, so a task may start as late as 23:45 and run its full
 * length into the small hours (`io/heuristic-placer.service.ts`).
 *
 * `prevStartMs`, when given, adds {@link stabilityScore}'s light penalty for
 * drifting away from the session's last manually-set start, so an otherwise
 * tied or near-tied candidate that's closer to it wins.
 */
export function bestFreeSlot(
  durationMinutes: number,
  occupied: Interval[],
  windowStart: Date,
  windowEnd: Date,
  prefMatrix: number[],
  timezone: string,
  fitWindowEnd: Date = windowEnd,
  prevStartMs?: number,
): Date | null {
  const durationMs = durationMinutes * MS_PER_MINUTE;
  // Both ceilings are typically capped to a session's deadline by the caller,
  // and deadlines are arbitrary user-entered instants — not guaranteed to be
  // slot-aligned. Round DOWN so a session is never placed to finish past the
  // exact deadline instant (rounding up here would let it finish up to one
  // slot late, which is an EDF-correctness bug).
  const startCeilMs = floorToSlot(windowEnd.getTime());
  const fitCeilMs = floorToSlot(fitWindowEnd.getTime());

  let best: { start: number; score: number } | null = null;
  for (
    let start = ceilToSlot(windowStart.getTime());
    start < startCeilMs;
    start += SLOT_MS
  ) {
    const end = start + durationMs;
    if (end > fitCeilMs) continue;
    if (overlapsAny(occupied, start, end)) continue;

    const total =
      slotPreferenceScore(prefMatrix, start, end, timezone) +
      (prevStartMs !== undefined ? stabilityScore(prevStartMs, start) : 0);
    if (best === null || total > best.score) {
      best = { start, score: total };
    }
  }

  return best ? new Date(best.start) : null;
}

File: services/bandit/src/core/slot_score.py
"""Slot scoring + heuristic best-free-slot with a vectorized scan
(port of ``slot-score.ts``)."""

from __future__ import annotations

import numpy as np
from numpy.typing import ArrayLike, NDArray

from .constants import (
    HOUR_MS,
    MS_PER_MINUTE,
    SLOT_MS,
    STABILITY_FAR_HOURS,
    STABILITY_NEAR_HOURS,
    STABILITY_SATURATION_HOURS,
    STABILITY_WEIGHT,
    STABILITY_WEIGHT_FAR,
    STABILITY_WEIGHT_NEAR,
    TIME_GRANULARITY,
)
from .preference import effective_preference_matrix
from .slot import Intervals, ceil_to_slot, floor_to_slot, local_cells

_TIE_DECIMALS = 9  # scores equal within 1e-9 tie -> earliest start wins


def stability_score(prev_start_ms: float, new_start_ms: float) -> float:
    dist_h = abs(new_start_ms - prev_start_ms) / HOUR_MS
    sat = min(dist_h, STABILITY_SATURATION_HOURS) / STABILITY_SATURATION_HOURS
    return -STABILITY_WEIGHT * sat


def stability_scores(
    prev_start_ms: float, starts_ms: NDArray[np.int64]
) -> NDArray[np.float64]:
    dist_h = np.abs(starts_ms - prev_start_ms) / HOUR_MS
    sat = np.minimum(dist_h, STABILITY_SATURATION_HOURS) / STABILITY_SATURATION_HOURS
    return np.asarray(-STABILITY_WEIGHT * sat, dtype=np.float64)


def stability_weight(prev_start_ms: float, now_ms: float) -> float:
    """Stability weight for a task whose current start is ``prev_start_ms``:
    ``NEAR`` while it is <= ``NEAR_HOURS`` away (or already past), fading
    linearly to ``FAR`` at ``FAR_HOURS``."""
    lead_h = max(0.0, prev_start_ms - now_ms) / HOUR_MS
    t = (lead_h - STABILITY_NEAR_HOURS) / (STABILITY_FAR_HOURS - STABILITY_NEAR_HOURS)
    t = min(1.0, max(0.0, t))
    return STABILITY_WEIGHT_NEAR + (STABILITY_WEIGHT_FAR - STABILITY_WEIGHT_NEAR) * t


def proximity_stability_scores(
    prev_start_ms: float, starts_ms: NDArray[np.int64], now_ms: float
) -> NDArray[np.float64]:
    """:func:`stability_scores` shape (saturating at ``SATURATION_HOURS``),
    scaled by :func:`stability_weight` instead of the fixed heuristic weight."""
    dist_h = np.abs(starts_ms - prev_start_ms) / HOUR_MS
    sat = np.minimum(dist_h, STABILITY_SATURATION_HOURS) / STABILITY_SATURATION_HOURS
    w = stability_weight(prev_start_ms, now_ms)
    return np.asarray(-w * sat, dtype=np.float64)


def slot_preference_scores(
    pref_matrix: ArrayLike,
    starts_ms: NDArray[np.int64],
    duration_minutes: int,
    timezone: str,
) -> NDArray[np.float64]:
    """Overlap-weighted preference score of ``[s, s + duration)`` for every
    slot-aligned start in ``starts_ms`` (one prefix-sum pass)."""
    if duration_minutes <= 0 or duration_minutes % TIME_GRANULARITY:
        raise ValueError("duration must be a positive multiple of 15")
    if starts_ms.size == 0:
        return np.empty(0)
    matrix = effective_preference_matrix(pref_matrix)
    n = duration_minutes // TIME_GRANULARITY
    first = int(starts_ms.min() // SLOT_MS)
    last = int(starts_ms.max() // SLOT_MS)
    span = last - first + n
    piece = matrix[local_cells(first, span, timezone)] * 0.25
    cs = np.concatenate(([0.0], np.cumsum(piece)))
    idx = (starts_ms // SLOT_MS - first).astype(np.int64)
    return np.asarray(cs[idx + n] - cs[idx], dtype=np.float64)


def slot_preference_score(
    pref_matrix: ArrayLike,
    start_ms: int,
    end_ms: int,
    timezone: str,
) -> float:
    dur = (end_ms - start_ms) // MS_PER_MINUTE
    return float(
        slot_preference_scores(
            pref_matrix, np.array([start_ms], dtype=np.int64), dur, timezone
        )[0]
    )


def free_start_mask(
    first_ms: int, n: int, duration_ms: int, occupied: Intervals
) -> NDArray[np.bool_]:
    """For the ``n`` slot-aligned starts ``first_ms + i*SLOT_MS``: True where
    ``[s, s + duration)`` overlaps no occupied interval (difference-array pass)."""
    diff = np.zeros(n + 1, dtype=np.int64)
    for o_start, o_end in occupied:
        # blocked starts: o_start - duration < s < o_end
        lo = (o_start - duration_ms - first_ms) // SLOT_MS + 1
        hi = -((first_ms - o_end) // SLOT_MS)  # ceil((o_end - first) / SLOT)
        lo, hi = max(lo, 0), min(hi, n)
        if lo < hi:
            diff[lo] += 1
            diff[hi] -= 1
    return np.asarray(np.cumsum(diff[:n]) == 0, dtype=np.bool_)


def best_free_slot(
    duration_minutes: int,
    occupied: Intervals,
    window_start_ms: int,
    window_end_ms: int,
    pref_matrix: ArrayLike,
    timezone: str,
    fit_window_end_ms: int | None = None,
    prev_start_ms: int | None = None,
) -> int | None:
    """Free 15-min-aligned start with the highest preference score (earliest
    wins ties); ``None`` when nothing fits."""
    duration_ms = duration_minutes * MS_PER_MINUTE
    start_ceil = floor_to_slot(window_end_ms)
    fit_ceil = floor_to_slot(
        window_end_ms if fit_window_end_ms is None else fit_window_end_ms
    )
    first = ceil_to_slot(window_start_ms)
    hi = min(start_ceil, fit_ceil - duration_ms + 1)  # exclusive upper bound
    if hi <= first:
        return None
    starts = np.arange(first, hi, SLOT_MS, dtype=np.int64)
    starts = starts[free_start_mask(first, starts.size, duration_ms, occupied)]
    if starts.size == 0:
        return None
    total = slot_preference_scores(pref_matrix, starts, duration_minutes, timezone)
    if prev_start_ms is not None:
        total = total + stability_scores(prev_start_ms, starts)
    total = np.round(total, _TIE_DECIMALS)
    return int(starts[int(np.argmax(total))])  # argmax -> first (earliest) max
